Improve `Dictionary::Get` perfomance

mike-f1 opened the pull request:
**Performance:** This PR improve performance of `Dictionary` by using [FNV1a](https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function) hashing funtion

## Summary

As outlined in #6329, a lot of time is spent in `Dictionary::Get` which is called hundreds of time per frame.

In order to lower the time spent there, instead of doing a full string comparison the strings are hashed and only these hash are compared.

This alone leads to a slight performance boost (around ~3%) because hashes are stored _inside_ the vector and there's no need to fetch strings from memory, making the hash computation comparable to the time needed to fetch those strings from (a "colder") memory.

But the bulk of the change is that the hashing algorithm is 'constexpr'-essed: this means that hashes can (and should) be computed at compile-time whenever it's possible, removing needs to hash strings at runtime.
Having a string hash directly computed at compile time means that at runtime the hash is directly used in the binary search inside the vector, reducing time by an average of 20/30% (though peak perfomance halve that time).

The compile time hashing of strings is obtained by simply adding the suffix `_fnv1a` to every string literal passed to `Dictionary::Get`, which cause the selection of an overload that use directly the hash to perform the lookup.

## Couple of Notes

* `stringAndHash` is needed because the _name_ (= the string) of an attribute is still needed: I think that storing the pointer to the char on the same structure is the easier way to retrieve these strings
* An hash collision may happens: I provide a [runtime-check](https://github.com/endless-sky/endless-sky/commit/a721a36feb9ce38e638408acd05e10722ebbb1e2) which trigger an exception if this happens because I think it's a good way to report a collision, but I'm open to changes.
* I improved runtime hashing string literals for only when players are in the game screen, I think further PR can improve loading times and performance in the various screen (starport, outfitter etc...)

## Save File
N/A

## Testing Done

The main test is a bit unusual, but provide a big number of samples:
the [first commit](https://github.com/endless-sky/endless-sky/commit/c3c0a781ddd71b5d038fe3755bed2948a7e94613) instruments the `Dictionary::Get` method using `std::chrono` and print the time averaged over 5 thousand samples.

In [that commit ](https://github.com/endless-sky/endless-sky/commit/91c4ec07dab1ac398a58f34da52be4f0c25bb7e1) the overload use the same method to provide profiling info.

The [last commit](https://github.com/endless-sky/endless-sky/commit/beb2bc7fb73b1f9f6fa72b6915b14bbc0e2f8d2e) revert the profiling.

Here some result obtained on my (really old laptop) machine:
[Dictionary_Get_Timing.ods](https://github.com/endless-sky/endless-sky/files/10068964/Dictionary_Get_Timing.ods)

## Additional Testing

* Tested that sorted insertion of elements by their hash keep vector sorted
* Tested that there aren't collision with the given set of strings

## Performance Impact

Reduction of 20/30% of times spent in `Dictionary::Get` for every compile-time string literal (peaking at ~50%), 
and a reduction of ~3% for run-time strings.

Hurleveur commented:
This really sounds promising, thx for working on it!
Dictionary calls are indeed super common.

mike-f1 commented:
@Hurleveur , thx's

Sorry for the force push: I changed `Ship.cpp` so it passes coding style and (btw) swapped a couple of commits
so they are more logical; I'm still waiting the next CI to see if I catch everything in Dictionary.

quyykk commented:
> swapped a couple of commits

just fyi we always squash merge into master, so you don't need to keep the commits clean. (also this looks really cool!)

mike-f1 commented:
> just fyi we always squash merge into master, so you don't need to keep the commits clean.

Well, just a matter of habits (although I feel it could improve 'reviewability')

> (also this looks really cool!)

Thank's again

QUESTION: I see code style checks are addressed calling `python ./utils/check_code_style.py` on CI but I'm unable to find that file on my local repo, where I can grab that file? Thx's in advance

quyykk commented:
> I see code style checks are addressed calling python ./utils/check_code_style.py on CI but I'm unable to find that file on my local repo, where I can grab that file? Thx's in advance

That file was added pretty recently, so you'll have to merge the upstream commits into your branch.

RisingLeaf commented:
I like it, it is a common method to reduce cpu usage.
Nice that someone is doing it

mike-f1 commented:
@RisingLeaf thx's, however the code for [FNV-1a](https://github.com/endless-sky/endless-sky/pull/7685/commits/8fbcab115e21abff85ce7865d40b426035243c60) came from [this old gist](https://gist.github.com/filsinger/1255697/1972eb676b47116838edaacf923e60b9173199c2), and it has been quite handy :smiley: : So I don't think I deserve all the glory...

RisingLeaf commented:
> @RisingLeaf thx's, however the code for [FNV-1a](https://github.com/endless-sky/endless-sky/pull/7685/commits/8fbcab115e21abff85ce7865d40b426035243c60) came from [this old gist](https://gist.github.com/filsinger/1255697/1972eb676b47116838edaacf923e60b9173199c2), and it has been quite handy 😃 : So I don't think I deserve all the glory...

You get the glory because you opened the PR

mike-f1 commented:
Just fixing style commits (I know, I'm a bit pedantic :sweat_smile: ) and adding one more because Big Boys told us that copying small objects is better then passing const references.

quyykk commented on source/Dictionary.cpp:
```diff
@@ -92,3 +103,17 @@ double Dictionary::Get(const string &key) const
+		return a.first.GetHash().Get() == b.first.GetHash().Get();
+	});
+	if(equal_found != (*this).end())
+		throw runtime_error { "Found an hash collision on '" + string((*equal_found).first.GetString()) + "'" };
```
Although hash collisions using FNV1A are pretty rare, they do happen and it seems less than ideal to crash the game if they happen. The usual solution is to compare the keys, which would amount to a single strcmp per lookup. I wonder if that has a negligible impact of performance or not.

mike-f1 replied:
For this function to work well, it should be called once a `Dictionary` has been completely filled.
That's the moment where a single call can be made for this kind of checks.
Indeed, after a (shallow) check over the codebase I recognize that Outfit initialization could be the right place to do that call just because it uses `operator[]` and I placed that call just after... But I don't know enough about the codebase (maybe there's a `Dictionary` for ship attribute and one for outfit? I don't know...).

> The usual solution is to compare the keys, which would amount to a single strcmp per lookup. I wonder if that has a negligible impact of performance or not.

I think it has a big impact :P

> it seems less than ideal to crash the game if they happen

:joy: You're right, but the 'reaction' (crashing game through an exception) although exaggerated is good one's: when a collision happens, it is better to not pass any test and report back to programmers as soon as possible, so an action can be taken (changing the seed could suffice).

So the point is: how you would me to report that a collision happened (excluding checks at compile time, they can't be done...) as soon as possible?

quyykk replied:
> Indeed, after a (shallow) check over the codebase I recognize that Outfit initialization could be the right place to do that call just because it uses operator[] and I placed that call just after... But I don't know enough about the codebase (maybe there's a Dictionary for ship attribute and one for outfit? I don't know...).

Yes that is fine. Ship attributes are actually stored internally as an Outfit, and an outfit's attributes are all parsed when Load is done. However, you could still have collisions with an attribute that we check for that isn't present on the outfit. For example, if `"hyperdrive"` and `"abc"` collide and there's an outfit just with `"abc"`, this function will report no problems although at runtime when the game checks for the hyperdrive attribute it will erroneously get `"abc"`.

 >  is better to not pass any test and report back to programmers as soon as possible

That's true, but attributes can be defined by plugin creators as well. So for them this is not exactly a very understandable problem.

mike-f1 replied:
> Yes that is fine. Ship attributes are actually stored internally as an Outfit, and an outfit's attributes are all parsed when Load is done. However, you could still have collisions with an attribute that we check for that isn't present on the outfit. For example, if "hyperdrive" and "abc" collide and there's an outfit just with "abc", this function will report no problems although at runtime when the game checks for the hyperdrive attribute it will erroneously get "abc".

I will check if there are remnant calls to `Dictionary::operator[]` (which is the only method allowed to insert new keys), but if you say that all attributes are loaded once Load is done, wouldn't it be worth to create a `Dictionary` just to put all of the loaded keys and check?

This will also reduce the number of calls to `Dictionary::CheckCollisions()`...

Where should I begin?

mike-f1 replied:
> That's true, but attributes can be defined by plugin creators as well. So for them this is not exactly a very understandable problem.

In this case, wouldn't it be good to fail on load (almost a CTD, I know...) but printing or notifying the plugin creator?

mike-f1 replied:
I think I found the right way, let me know if that's fine for you.

quyykk commented on source/Dictionary.h:
```diff
@@ -20,25 +20,106 @@ this program. If not, see <https://www.gnu.org/licenses/>.
+{
+	using def_type = uint32_t;
+
+	template <typename S> struct fnv_internal;
+	template <typename S> struct fnv1;
+	template <typename S> struct fnv1a;
+
+	template <> struct fnv_internal<def_type>
+	{
+		constexpr static def_type default_offset_basis = 0x811C9DC5;
+		constexpr static def_type prime = 0x01000193;
+	};
+
+	template <> struct fnv1<def_type> : public fnv_internal<def_type>
+	{
+		constexpr static inline def_type hash(char const *const aString, const def_type val = default_offset_basis)
+		{
+			return (aString[0] == '\0') ? val : hash(&aString[1], (val * prime) ^ def_type(aString[0]));
+		}
+	};
+
+	template <> struct fnv1a<def_type> : public fnv_internal<def_type>
+	{
+		constexpr static inline def_type hash(char const *const aString, const def_type val = default_offset_basis)
+		{
+			return (aString[0] == '\0') ? val : hash(&aString[1], (val ^ def_type(aString[0])) * prime);
+		}
+	};
```
You can simplify this massively:

1. Why are the functions inside a struct?
2. `fnv1` is unused, since you're using `fnv1a` only
3. To avoid having the constants in global namespace, I think it's fine if you inline them, since they're only used once and are fairly opaque anyways.

mike-f1 replied:
Yes, you're right: I only waited to know if the algorithm I choose is ok...
At that time I take the first one I saw :grin:, which reveals a discrete luck as this one still quite fast and really simple...

mike-f1 commented on source/Dictionary.h:
```diff
@@ -20,25 +20,106 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
+// Code from:
+// https://gist.github.com/filsinger/1255697/1972eb676b47116838edaacf923e60b9173199c2
+namespace hash_fnv
```
After the massive simplification you spotted (it needs only a single algorithm, so a lot less LOC), do you keep thinking it should go in a new file?
If so, would `fnv1a.h` be good?
And would you put that file in source/ directly or in a sub-directory?

quyykk replied:
> After the massive simplification you spotted (it needs only a single algorithm, so a lot less LOC), do you keep thinking it should go in a new file?

Yeah, to keep things organized I guess.

>  If so, would fnv1a.h be good?

sounds good

>  And would you put that file in source/ directly or in a sub-directory?

just put it in source yeah

mike-f1 commented on source/Dictionary.h:
```diff
@@ -20,25 +20,106 @@ this program. If not, see <https://www.gnu.org/licenses/>.
+};
+
+
+inline constexpr HashWrapper operator "" _fnv1a (const char *aString, size_t aStrlen)
```
Ok, going for `_hash`!

mike-f1 commented:
[OT] BTW, I have 2 problems. I post here as I don't know if these are common or I'm alone.
1)  `utils/check_code_style.py` give errors:
```
Traceback (most recent call last):
  File "utils/check_code_style.py", line 576, in <module>
    (e, w) = check_code_style(file, contents)
  File "utils/check_code_style.py", line 164, in check_code_style
    lines = [line.removesuffix('\n').removesuffix('\r') for line in lines]
  File "utils/check_code_style.py", line 164, in <listcomp>
    lines = [line.removesuffix('\n').removesuffix('\r') for line in lines]
AttributeError: 'str' object has no attribute 'removesuffix'

```
`python --version` give `Python 3.8.10`
and `pip list` output includes `regex                        2019.8.19`
...so I'm stuck :cold_sweat:

2) How can I run unit test? ...There's a page on wiki I didn't find?

quyykk commented:
>  How can I run unit test? ...There's a page on wiki I didn't find?

Hmmm I'm not sure there is documentation now that you mention it. `scons test`. If you want to run the benchmarks: `./tests/unit/endless-sky [!benchmark]`

tibetiroka commented:
It runs fine on Python 3.10.6. Any chance you could upgrade to that?

mike-f1 commented:
> It runs fine on Python 3.10.6. Any chance you could upgrade to that?

I will try, but not immediately: Let me prepare mentally to take on an upgrade :sweat_smile:

Amazinite commented on source/AI.cpp:
```diff
@@ -197,7 +197,7 @@ namespace {
 	{
 		if(!fuelCapacity)
-			fuelCapacity = ship.Attributes().Get("fuel capacity");
+			fuelCapacity = ship.Attributes().Get("fuel capacity"_fnv1a);
```
Is it really necessary to have to append `_fnv1a` to every attribute name that we want to access now? Can the hash part not be something that `Get` handles?

mike-f1 replied:
I fear yes: appending "_fnv1a" means you are calling a function (`operator""`) and that member needs also to be called at compile time.

Put another way: the `Get` method is receiving an `HashWrapper` as a parameter, not a `char *`.

But if anyone know about a workaround I will do.

mike-f1 replied:
BTW: As per comment review, I'm changing "_fnv1a" -> "_hash"

mike-f1 replied:
@Amazinite; @Hurleveur; @quyykk : I'm realizing I'm introducing something a bit new in this codebase, so are you all fine with this or do you prefer me to stick on "standard", eg:

`ship.Attributes().Get(HashOf("fuel capacity"));`

petervdmeer replied:
I prefer using the "standard" HashOf(); it seems more readable/more clear.

mike-f1 commented:
Added commit [aeb9104](https://github.com/endless-sky/endless-sky/pull/7685/commits/aeb9104a26d005a7b1afc8bc22bdfcf039cd2f9d), which should resolve the issues raised by @quyykk 
1) I found a spot to place a check on the whole set defined attributes: that's in `UniverseObjects` when
checks on names of every `Outfit` are made.

2) In that spot it seems throwing exceptions is handled; indeed, raising a 'fake collision' will give this:
![Found_A_Fake_Collision](https://user-images.githubusercontent.com/13293916/203960255-a15b4047-6b52-4fe5-a417-1063d87034d4.jpeg)

petervdmeer commented:
>     1. I found a spot to place a check on the whole set defined attributes: that's in `UniverseObjects` when
>        checks on names of every `Outfit` are made.
> 
>     2. In that spot it seems throwing exceptions is handled; indeed, raising a 'fake collision' will give this:

I really like the use of hashes instead of the original strings to improve performance, but the error-message "Found a hash collision, please try a different key" is totally unacceptable; we are not going to forbid the use of certain keys because our algorithm cannot cope with it. Our algorithm then just needs to be better.

petervdmeer commented:
Note; PRs https://github.com/endless-sky/endless-sky/pull/7165 and https://github.com/endless-sky/endless-sky/pull/5748 can be considered variants of this PR, but where the hashing algorithm gets replaced by array-indexes (in case of 7165) and pointers (in case of 5748).
The benefit of this PR is that the strings hardcoded in the game can be resolved to a hash _at compile time_, but most of the strings we deal with come from data-files and save-games, so they still need to be interned.
Both of the other PRs will not have any collisions, because they generate the unique identifiers at load time without a mechanism that would map 2 different strings to the same identifier.

mike-f1 commented:
I'm sorry for the [TL;DR] answer, but there are a lot of topics to talk about...

> I really like the use of hashes instead of the original strings to improve performance, but the error-message "Found a hash collision, please try a different key" is totally unacceptable; we are not going to forbid the use of certain keys because our algorithm cannot cope with it. Our algorithm then just needs to be better.

Hash algorithm can collide: that's the truth for every hashing algorithm.
However, given the number of the actual (and future?) keys over the number of combinations of an `uint32_t`, it's very (very!) unlikely it happens.
But, if it's going to happens, then "all bets are off" and that's the reason behind my effort to check if that happens.

The action to take can be easy: changing the new key (even slightly) or changing the seed (*) can be enough, or changing the algorithm.
That what's behind the part "or contact developers": when (if?) a collision happens, then someone can change the seeds and verify that the newly added key (and all those already there) aren't colliding.

I'm open to proposal on how to improve that feedback, and I'm still wondering if that is the right place, which should catch every new insertion.

> but most of the strings we deal with come from data-files and save-games, so they still need to be interned

That's true, and there's is no way around: indeed both PRs you cited are still "interning" strings because these need to be there when you save or show a key.
However, as I stated in PR description, I didn't touch code for loading. But it's trivial to change and can be quite effective: If you load the game [dropping this commit](https://github.com/endless-sky/endless-sky/pull/7685/commits/bddeb2b9bd48e70d3c8c89c5325a4f2892ae0182), you can see that `Get()` is called more than 15'000 times.

> Both of the other PRs will not have any collisions, because they generate the unique identifiers at load time without a mechanism that would map 2 different strings to the same identifier.

I would repeat: that's very unlikely to happens, eg the [Java have a "general purpose" algorithm](https://www.w3schools.com/java/ref_string_hashcode.asp) for string which is even simpler than this.

Anyway, even with this there's needs to perform a binary search over the hashes that, even if faster, can't compete with a direct access over a vector (which bypass that binary search).
I would point I think that that access is likely not be "massively" faster: `Dictionary::Get` is called "sparsely" and I think is memory bounded.

However, a big difference I see is on how at call site things are handled: in #7165 it's not (yet?) shown how `Get(const dict &key)` is called (it is called nowhere actually), and I highly doubt it's an "easy" change: you (as a programmer) should memorize every `dict`, or rely on `#define`s which lead to a known overkill.
Having compile time hashes, either by using `_something` or using `HashOf`, is a simple change: indeed (but tell me if I miss something), I _effectively_ changed performance, while in #7165 there's only a "promise" to improve performance (that PR seems to lack changes on call sites).

I don't understand if/how #5748 deals with the exact same problem, I feel it's working on a different problem... But probably I miss something...
Anyway, in this PR I didn't work on the 'Interner', as it's usage become less and less crucial with the progress on using hashes.
Just a thought: I think it's perfomance can be improved by using a single `std::string` (so a single buffer in which you add every string not yet stored) and using a "sort of `string_view`" (this is C++11...) for indexing (you have to deal with relocations of the buffer, so probably you should store offsets).

NOTE *: @quyykk, I left these constants out of the algorithm: they are inside a struct which is inside a namespace, so I doubt they are easily accessible from outside (they are also constexpr...), let me know if you're fine with that.

mike-f1 commented:
It seems something [is going wrong](https://github.com/endless-sky/endless-sky/actions/runs/3547675654/jobs/5969999386), but I don't understand what exactly...

petervdmeer commented:
> Hash algorithm can collide: that's the truth for every hashing algorithm. However, given the number of the actual (and future?) keys over the number of combinations of an `uint32_t`, it's very (very!) unlikely it happens.

I initially agreed on this one, but then decided to run some numbers with an "online Hash Collision Calculator" to get some actual numbers for "very unlikely"; The chance of having 2 items map to the same 32-bit hash-value is around 1% when we get to 10000 different strings, so that sounds somewhat reasonable. But if we get to 100000 different strings (for whatever reason), then the probability of having _any_ collision increases to 69%, at which point just changing the seed is likely resulting in a new collision on two other strings.

> But, if it's going to happens, then "all bets are off" and that's the reason behind my effort to check if that happens.

This is the part that I don't like. Hashes are very useful for highly efficient storage, comparisons and data-retrievals, but my experience is that systems that use hashes for optimizations do actually also ensure that collisions are handled gracefully; still showing the correct behaviour on colissions (with reduced performance on the items that collide).

> The action to take can be easy: changing the new key (even slightly)

No, the keys are provided by content creators, they are "part of lore" for the game and shown in the user-interface. This suggestion is similar to suggesting that we introduce small spelling errors in some texts because our renderers cannot always process the correct spelling. Content creators should just not be restricted in the keys they are allowed to choose, in the same way that we actively drive for correct (US/American) spelling in all the texts that we use in-game.

> or changing the seed (*) can be enough

Maybe when we stay below 10000 strings, no if we get to 100000 strings.

>, or changing the algorithm.

Maybe, but as you said; "Hash algorithm can collide: that's the truth for every hashing algorithm.", so we always need to be prepared to deal with collisions, regardless of which hashing algorithm we use.

> That what's behind the part "or contact developers": when (if?) a collision happens, then someone can change the seeds and verify that the newly added key (and all those already there) aren't colliding.

Detecting a collision and then just telling a content-creator to contact a developer is not a graceful way to handle the collision. End-users that run multiple plug-ins might also run into issues with collisions in-between plug-ins, which will make a very rare issue actually very difficult to analyse and debug.

My PR https://github.com/endless-sky/endless-sky/pull/5583 is currently blocked on something ungraceful (LTO, not relevant to this PR), and I'm sufficiently senior within the project to actually deal with this, but I don't expect someone that just deals with a plug-in to properly route a hash collision to a developer. (Especially when most developers only test on the vanilla game and would thus not be able to reproduce the issue.)

> I'm open to proposal on how to improve that feedback, and I'm still wondering if that is the right place to call `AddKeysWhileChecking`, which should catch every new insertion.

I have the feeling that AddKeysWhileChecking will only catch the "runtime generated keys that are interned", or is there also a way to check for collisions with the "hashes that are generated compile-time from the hardcoded given strings"?

> > Both of the other PRs will not have any collisions, because they generate the unique identifiers at load time without a mechanism that would map 2 different strings to the same identifier.
> 
> I would repeat: that's very unlikely to happens, eg the [Java have a "general purpose" algorithm](https://www.w3schools.com/java/ref_string_hashcode.asp) for string which is even simpler than this.

I'm sure that Java uses the hash wisely and gracefully deals with collisions at locations where those matter (and they use 32-bit hashes, so the numbers should be similar to what I found above).

> However, a big difference I see is on how at call site things are handled: in #7165 it's not (yet?) shown how `Get(const dict &key)` is called (it is called nowhere actually), and I highly doubt it's an "easy" change: you (as a programmer) should memorize every `dict`, or rely on `#define`s which lead to a known overkill. Having compile time hashes, either by using `_something` or using `HashOf`, is a simple change: indeed (but tell me if I miss something), I _effectively_ changed performance, while in #7165 there's only a "promise" to improve performance (that PR seems to lack changes on call sites).

For #7165 I expect that the relevant change is to use `Intern`, similar to how `HashOf` is used in this PR; just make sure that you always compare "interned references"; the interning process itself handles de-duplication, so two "interned references" that are equal are pointing to the same value.

> I don't understand if/how #5748 deals with the exact same problem, I feel it's working on a different problem... But probably I miss something...

#5748 aims at replacing the string-comparisons with char* comparisons, similar to how this PR aims at replacing the string-comparisons with hash-value-comparisons. (But it is my mistake that this information was missed; I didn't clearly communicate this and I only added the mechanism without actually using it.)

> Anyway, in this PR I didn't work on the 'Interner', as it's usage become less and less crucial with the progress on using hashes.

The interner would indeed no longer be needed when we just perform hash-based comparisons (except maybe for detecting the hash-collisions?).

> Just a thought: I think it's perfomance can be improved by using a single `std::string` (so a single buffer in which you add every string not yet stored) and using a "sort of `string_view`" (this is C++11...) for indexing (you have to deal with relocations of the buffer, so probably you should store offsets).

This sounds like something which we better keep for a separate PR (outside of this PR).

mike-f1 commented:
> I initially agreed on this one, but then decided to run some numbers with an "online Hash Collision Calculator" to get some actual numbers for "very unlikely"; The chance of having 2 items map to the same 32-bit hash-value is around 1% when we get to 10000 different strings, so that sounds somewhat reasonable. But if we get to 100000 different strings (for whatever reason), then the probability of having any collision increases to 69%, at which point just changing the seed is likely resulting in a new collision on two other strings.

...Well, probably my fault underestimating the problem because I saw that actual code base deals with [170 strings](https://github.com/endless-sky/endless-sky/pull/7165#issue-1348874284) or similar numbers, which is far from the numbers you use. But I think that if you use that site you can tell me:
* the probability of a collision with 200 strings (a bit of margin above the actual set)
* the probability of a collision using a 64-bit hashing algorithm over 100'000; wouldn't it lower the probability so it still reasonable?

...And if not then what about a 128-bit or 256-bit hashing... Because hashing scale well :sunglasses:.

> This is the part that I don't like. Hashes are very useful for highly efficient storage, comparisons and data-retrievals, but my experience is that systems that use hashes for optimizations do actually also ensure that collisions are handled gracefully; still showing the correct behaviour on colissions

That's also my fault: I didn't understand what's the problem you speak about..
There's something about ["collision mitigation"](https://en.wikipedia.org/wiki/Hash_collision#Collision_resolution) which can be implemented, simply I didn't because I though it's an overkill.
But if that's is considered necessary, I can give it a try.

> (with reduced performance on the items that collide).

Indeed: even with a 'Collision Mitigation' there still a need to give some feedback: you wouldn't have a lot of collisions (without knowing about) because they slow down the game.

> Maybe when we stay below 10000 strings, no if we get to 100000 strings.

This game date back 4/5 years (?) and in these years the number of attributes grown but still less than 200, so how you would reach 10'000 or 100'000 strings (starting from less than 200)? And why?

> I have the feeling that AddKeysWhileChecking will only catch the "runtime generated keys that are interned"

...which is what it should check: or rephrased better "it catch the runtime string and hashes of the keys that are interned"

> or is there also a way to check for collisions with the "hashes that are generated compile-time from the hardcoded given strings"?

Aren't they the same? I check for those loaded, and obviously those used in code aren't colliding because they are the same.

...And maybe there's a way to check at compile time if strings in code are correct, so you can't compile "sheld" (which is a typo), just wondering...

> I'm sure that Java uses the hash wisely and gracefully deals with collisions at locations where those matter (and they use 32-bit hashes, so the numbers should be similar to what I found above).

No, [definitely no mitigation](https://stackoverflow.com/questions/9406775/why-does-string-hashcode-in-java-have-many-conflicts) :sweat: , but an interesting read btw 

> For #7165 I expect that the relevant change is to use Intern, similar to how HashOf is used in this PR; just make sure that you always compare "interned references"; the interning process itself handles de-duplication, so two "interned references" that are equal are pointing to the same value.

How you can compare an "interned reference" (likely a `char *` pointing to heap, I suppose) with a pointer to a string literal (which is instead stored in program segment)? That's what is puzzingly me: what about call sites?

Asking because I think that a point on "plus" of this PR is that the call sites are easy to change (`something.Get("whatever")` to `something.Get("whatever"_hash)`, or `HashOf` of course), and I didn't see how #7165 changes all those call sites.

Could you elaborate a bit on this? Showing a call site after #7165?

> #5748 aims at replacing the string-comparisons with char* comparisons, similar to how this PR aims at replacing the string-comparisons with hash-value-comparisons

I can't see that, sorry... Here I replace both `char *` and `string` comparisons, but I didn't see differences between actual `operator[]` implementation and `StringInterner::Intern` implementation.

Btw, I see [there is](https://github.com/endless-sky/endless-sky/pull/5748/files#diff-b1d48ed7cbb4f374f89cbb6d8ab874d603618f82807a5c52fa97f4cf3438fc05) no more a read-only `Get` method; wouldn't it be worst than actual because `StringInterner::Intern` is always using a mutex?

> The interner would indeed no longer be needed when we just perform hash-based comparisons .

Nope: the `Interner` is still needed especially using hashes because you should handle the string to those classes that need to _show_ the 'key', but that's not performance critical and I left it as is.

> (except maybe for detecting the hash-collisions?)

That's how it worked, then I added `AddKeysWhileChecking` in 5df106b7

> This sounds like something which we better keep for a separate PR (outside of this PR).

Indeed (and if there are plans to keep FPS above 1 with 10'000~100'000 keys, that's the second _ most important thing to do_, right now).

So, just to recap:
I used FNV-1a thinking at ~200 strings, but if plans are to increase this number to 10'000 I can change the algorithm (...or this can be postponed to when someone will raise its hand telling "I can't use plugin X and plugin Y because game crash telling me that 'whatever key A' and 'whatever key B' are colliding", which is _quite_ easy to solve at that point: fake an outfit with those 2 keys, change seed, algo... whatever and see if game reach main screen; or use a [collision mitigation](https://en.wikipedia.org/wiki/Hash_collision#Collision_resolution)).

However, @petervdmeer are right: dealing with 10'000 or 100'000 strings means that everything with a binary search become too slow. What I fear is that _even_ a direct access could become slow with these numbers (and can introduce a lot of maintenance burden) for something I don't know about and I have hard time understanding: how and why you would put 100'000 keys there?

petervdmeer commented:
>     * the probability of a collision with 200 strings (a bit of margin above the actual set)
>     * the probability of a collision using a 64-bit hashing algorithm over 100'000; wouldn't it lower the probability so it still reasonable?

Both seem quite reasonable indeed, I don't want to endorse any particular site, but the one I used is https://kevingal.com/apps/collision.html

> ...And if not then what about a 128-bit or 256-bit hashing... Because hashing scale well.

Yes, this seems to scale well, just adding a single bit doubles the amount of hashing space. But when we make the hashes too large, then we will somewhere get to a point where just performing the string compare (which can compare 8 characters using a single 64-bit compare) becomes the more performant option. :smiley: 

> There's something about ["collision mitigation"](https://en.wikipedia.org/wiki/Hash_collision#Collision_resolution) which can be implemented, simply I didn't because I though it's an overkill. But if that's is considered necessary, I can give it a try.

I'm usually very much a fan of risk-avoidance; but the chances of something going wrong already seems very small with 64-bit hashes.

> > Maybe when we stay below 10000 strings, no if we get to 100000 strings.
> 
> This game date back 4/5 years (?) and in these years the number of attributes grown but still less than 200, so how you would reach 10'000 or 100'000 strings (starting from less than 200)? And why?

Automatic conditions were recently introduced, and they are rapidly becoming popular. If we ever get to real-time automatic conditions, then we might want to re-use the hashing-solution that was chosen here.

> > I have the feeling that AddKeysWhileChecking will only catch the "runtime generated keys that are interned"
> 
> ...which is what it should check: or rephrased better "it catch the runtime string and hashes of the keys that are interned"

So we agree on this one.

> > or is there also a way to check for collisions with the "hashes that are generated compile-time from the hardcoded given strings"?
> 
> Aren't they the same? I check for those loaded, and obviously those used in code aren't colliding because they are the same.

We typically write tests for a subset of the attributes, so if two rarely used attributes map to the same hash, then the tests might not notice the collision (because those two attributes are never used together in the same test).

> ...And maybe there's a way to check at compile time if strings in code are correct, so you can't compile "sheld" (which is a typo), just wondering...

I'm not worried about typos on the attributes. I'm worried about rarely used attributes that pass all our testing, and show up only in a very specific situation where one of our top content-contributors is seeing all kinds of strange issues due to combining some attributes in a story, without anybody else being in a position to easily reproduce the issue (because others just didn't get the new story content yet).

The game is mainly a story-telling engine, so we should make sure our top content-creators can create nice story content without getting hit by very hard to debug issues.

> > I'm sure that Java uses the hash wisely and gracefully deals with collisions at locations where those matter (and they use 32-bit hashes, so the numbers should be similar to what I found above).
> 
> No, [definitely no mitigation](https://stackoverflow.com/questions/9406775/why-does-string-hashcode-in-java-have-many-conflicts) sweat , but an interesting read btw

I don't expect the mitigations within the hashing-algorithm itself, but in the Java-core classes that use hashing for optimized performance.
I did already read this link before posting my earlier reply, but this looks like a developer that just want to use hashes without wanting to deal with collisions.

> > For #7165 I expect that the relevant change is to use Intern, similar to how HashOf is used in this PR; just make sure that you always compare "interned references"; the interning process itself handles de-duplication, so two "interned references" that are equal are pointing to the same value.
> 
> How you can compare an "interned reference" (likely a `char *` pointing to heap, I suppose) with a pointer to a string literal (which is instead stored in program segment)? That's what is puzzingly me: what about call sites?

One interns both the loaded data, and the hardcoded data. After interning both, we have char-pointers on both sides, and we can just compare the pointers for equality without needing to actually compare the strings (because each string is interned only once, so it should not happen that two different pointers actually point to the same string).

> Asking because I think that a point on "plus" of this PR is that the call sites are easy to change (`something.Get("whatever")` to `something.Get("whatever"_hash)`, or `HashOf` of course), and I didn't see how #7165 changes all those call sites.
> Could you elaborate a bit on this? Showing a call site after #7165?

`something.Get("whatever")` gets replaced by `something.Get(Intern("whatever"))` on all those call sites.

> > #5748 aims at replacing the string-comparisons with char* comparisons, similar to how this PR aims at replacing the string-comparisons with hash-value-comparisons
> 
> I can't see that, sorry... Here I replace both `char *` and `string` comparisons, but I didn't see differences between actual `operator[]` implementation and `StringInterner::Intern` implementation.

This PR replaces `char *` and `string` comparisons by hash comparisons, 5748 aims for making the `char *` pointers themselves comparable (without needing to look at the string that is pointed at), by just Interning on both sides. The interning process ensures that each unique string gets represented by only one `char *`.

> Btw, I see [there is](https://github.com/endless-sky/endless-sky/pull/5748/files#diff-b1d48ed7cbb4f374f89cbb6d8ab874d603618f82807a5c52fa97f4cf3438fc05) no more a read-only `Get` method; wouldn't it be worst than actual because `StringInterner::Intern` is always using a mutex?

[A multiple-readers, single-writer lock](https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock) would be a very good idea for PR #5748 (or a follow-up pr if #5748 gets merged), but I'm not sure if that is what you were referring to.

Please also feel free to post review-comments on #5748 directly if you see something there that would benefit from some feedback.

> > The interner would indeed no longer be needed when we just perform hash-based comparisons .
> 
> Nope: the `Interner` is still needed especially using hashes because you should handle the string to those classes that need to _show_ the 'key', but that's not performance critical and I left it as is.

Good, we agree on keeping this as-is for this PR.

> However, @petervdmeer are right: dealing with 10'000 or 100'000 strings means that everything with a binary search become too slow. What I fear is that _even_ a direct access could become slow with these numbers (and can introduce a lot of maintenance burden) for something I don't know about and I have hard time understanding: how and why you would put 100'000 keys there?

We will not put 100000 keys there for the Attributes, but we might want to use a similar mechanism for Conditions in the future and we will have more conditions than that we have attributes.
(And I honestly don't really know for sure how many conditions we will actually have in the future, but I choose 100000 keys so that the percentages that I got out of the calculator look somewhat impressive.)

mike-f1 commented:
First of all, thank's for the answer: at least now I know more pieces of the puzzle.

I don't know about 'automatic conditions' (do you mind to give a hint or a link?) but if they are topping at those numbers probably they could become a big expense... However `Dictionary::Get` is everywhere, I hope is not the same for those :sweat_smile:.

I don't understand exactly the way this should take, however even with this PR you still have the big problem of the binary search: this still, perfomance wise, the main issue of this PR if key number raise.

#7165 deals with that offering a direct access, so it's faster.

This seems the choice: if performance is mandatory, then you have to go straight to #7165, else if you can keep the binary search then this is affordable.

However, you should consider that #7165, without losing so much, can be rewrited and simplified down to:
`Dictionary`
```
OffsetWrapper GetOffsetOf(char *key) const {
	pair<size_t, bool> pos = Search(key, *this);
	return OffsetWrapper(pos.first);
}

double Get(OffsetWrapper ow) {
	return (*this)[ow.Get()].second;
}
```
Where `OffsetWrapper` can be an opaque class, more or less:
```
class OffsetWrapper {
	friend class Dictionary;
public:
	// Nobody can fiddle with me, only dear Dictionary :heart: 
private:
	size_t Get() const { return offset; }
	OffsetWrapper(size_t o) : offset(o) {}

	size_t offset;
};
```

Showing that both can live together.

I thought also about 2 (crazy) ideas to improve actual performance:
* Make an internal cache (so you can also stop doing that externally), like that used for string SBO (a `std::array` where you put say 20~30 keys).
* Split the vector so you have less keys to search for in each of them. 

And these two approaches can also be combined...

2 cents about the 'CTD for collision' (or why I keep not changing my mind :grin:):
C++ gurus are repeating often that compile time errors are better than runtime errors, and that's why I think is preferable a CTD than mitigation: you _sure_ get the feedback and take the right action to solve the problem, while not handling that thing means (sort of) UB of the game and mitigation means slowdowns...
Should sounds better than 

> I'm worried about rarely used attributes that pass all our testing, and show up only in a very specific situation

Of course, I can still give a try to mitigation. No idea if that will be easy or not...

Let me know what you wish me to do.

Zireael07 commented:
Hm, why not both? Direct access for the 10000 conditions or whatever and this for conversations where you won't get 10000 things to check?

mike-f1 commented:
> Hm, why not both? Direct access for the 10000 conditions or whatever and this for conversations where you won't get 10000 things to check?

I'm not sure I understand everything because I don't know about "conditions" and "conversations"...

But yes: I think the main point of my last TL;DR answer is that implementing the overload of `Get` method as I have outlined can skip binary search. This means only a single "jump" when you call that `Get`.
And if SBO-like structure is then implemented, calling that overload is like calling directly a value stored in Ship or in Outfit (indeed `Dictionary` is a member thus getting a value from its SBO will be just an offset from class address).

petervdmeer commented:
> First of all, thank's for the answer: at least now I know more pieces of the puzzle.

You're welcome. And also thank you for your response; I also learned some more pieces of the puzzle of the 3 involved PRs.

> I don't know about 'automatic conditions' (do you mind to give a hint or a link?) but if they are topping at those numbers probably they could become a big expense... However `Dictionary::Get` is everywhere, I hope is not the same for those

I recently reviewed [one example](https://github.com/endless-sky/endless-sky/pull/7828), [another example](https://github.com/endless-sky/endless-sky/pull/7797) and [one more example](https://github.com/endless-sky/endless-sky/pull/7791), and I didn't (yet) review [another one currently in progress](https://github.com/endless-sky/endless-sky/pull/7792).

Conditions are currently not used every frame in performance critical code, but we probably want to optimize them somewhere in the future. But we should probably just keep them out of scope of this PR; the Dictionary performance is currently way more important than the conditions-evaluation performance, because Dictionary comparisons happen every frame, and conditions are (currently) only evaluated on certain "game events" (like landing, boarding, etc).

> I don't understand exactly the way this should take, however even with this PR you still have the big problem of the binary search: this still, perfomance wise, the main issue of this PR if key number raise.> 

Let's for now assume that the number of attributes (the number of Dictionary entries) doesn't rise much. Sorry for confusing the discussion with auto conditions (which are not that much related).

> #7165 deals with that offering a direct access, so it's faster.
> This seems the choice: if performance is mandatory, then you have to go straight to #7165, else if you can keep the binary search then this is affordable.

Good feedback! I used it in https://github.com/endless-sky/endless-sky/pull/7165#discussion_r1045138427 because it also seems to have some undesirable memory-usage effects (if I understand correctly). I'm not sure if the additional memory usage also has some negative performance effects (like having the dictionaries no longer fit in processor caches).

> However, you should consider that #7165, without losing so much, can be rewrited and simplified down to: `Dictionary.h`

Yes, it stores offsets instead of strings.

> I thought also about 2 (crazy) ideas to improve actual performance:
>     * Make an internal cache (so you can also stop doing that externally), like that used for string SBO (a `std::array` where you put say 20~30 keys).
>     * Split the vector so you have less keys to search for in each of them.
> And these two approaches can also be combined...

Nice ideas, we might even want to split the Attributes(Dictionary) into "commonly used attributes" and "not commonly used attributes", where we optimize for the commonly used attributes, at the expense of the less commonly used attributes becoming slower.
SBO is Small Buffer Optimization? Also looks like a nice feature.

> 2 cents about the 'CTD for collision' (or why I keep not changing my mind grin): C++ gurus are repeating often that compile time errors are better than runtime errors, and that's why I think is preferable a CTD than mitigation: you _sure_ get the feedback and take the right action to solve the problem, while not handling that thing means (sort of) UB of the game and mitigation means slowdowns...

I fully agree that compile time error-messages are very preferable above run-time error messages.

> Should sounds better than
> 
> > I'm worried about rarely used attributes that pass all our testing, and show up only in a very specific situation

This worry of mine applies to your PR; that we get issues not at compile time, but at runtime at a PC of one of our content-writers that just doesn't know anything about C++ and string-hashing.
The attributes (in the Dictionary) are not only coming from the C++-codebase; the attributes are also coming from data-files.

> Of course, I can still give a try to mitigation. No idea if that will be easy or not...

If we can (somehow) make sure that hash-collisions only happen _at compile time_ and _on a PC/laptop/whatever owned by one of our C++ coders_, then that is sufficient mitigation to me.
But if hash-collisions can happen on a PC/laptop/whatever owned by one of our story writers, then my worry is not addressed yet. Here is an example of [some obscure attribute in an obscure plugin which nobody except the plugin-author ever heard about](https://github.com/petervdmeer/endless-sky-undisclosed-plugin/blob/cef74d536188abb8fcec4b15f8c77ef6ea48b91b/data/clusterers/clusterers_weapons.txt#L10).

Those attributes are not visible in the code anywhere, because they only get used to match ships to outfits.
Maybe we should move such "non-realtime attributes" to a different category, since they don't need to be queried at runtime, but they are attributes in the current codebase, and they still are attributes in the current state of your PR. So we need to have a story on how to deal with hash collisions on those attributes that are just not visible in the codebase at all.

> Let me know what you wish me to do.

Your active participation in the discussion is very valuable, please keep voicing your opinion and your ideas.
And if we can find a way to make sure that hash-collisions always lead to (C++)compile-time errors, then my concerns are handled.

> EDIT: I expect that that "Endless-Sky not-so-blue screen of death" would (with 32bit over say 1'000 keys or 64bit over say 100'000 keys) show up a time every a 3~5 year I think, but I would put a disclaimer here smirk "Software is given AS IS, no responsibility" grin

I know, those calculators assume a smooth distribution. Performance could be worse and we could also be unlucky, but we could also just be very lucky and never see issues.

mike-f1 commented:
> I recently reviewed https://github.com/endless-sky/endless-sky/pull/7828, https://github.com/endless-sky/endless-sky/pull/7797 and https://github.com/endless-sky/endless-sky/pull/7791, and I didn't (yet) review https://github.com/endless-sky/endless-sky/pull/7792.

...they are exploding more than growing :sweat_smile:

Well: I can't (yet) answer properly (...I'm reading the links above, it's going to take a bit :sweat: ) but I think these two `Dictionary`, that for attributes and that for conditions, have different constraints and because of that it is better to keep their _data_ separate.
Searching over ~500 is quite different from ~10000, and also the 'perfomance pressure' is different (thousand per frame vs only on landing). If needed, that class can be templated so one can use some 'trait' to instance the best `Dictionary` for its case.

I don't even know if "the `Dictionary` for conditions" can be identified and if it work the same way as `Dictionary`, however I think that both can use a better storage for strings (waiting for something even more advanced, be it hashes, SBO, a (damn) counter or something else) and (I hope...) it should be an easy improvement: indeed storing strings individually means you have them `new`ed around (first indirection) with some of their content allocated somewhere else due to not fitting in SBO (second indirection).

If it's useful for "Conditions" (and will be ported), I can check it works in `Dictionary` and a follow up PR can put the bridge between these two.

This can't get the same improvement of this PR, though...

TL;DR details:

> I recently reviewed 7828, 7797 and 7791, and I didn't (yet) review 7792.
Conditions are currently not used every frame in performance critical code, but we probably want to optimize them somewhere in the future. But we should probably just keep them out of scope of this PR; the Dictionary performance is currently way more important than the conditions-evaluation performance, because Dictionary comparisons happen every frame, and conditions are (currently) only evaluated on certain "game events" (like landing, boarding, etc).

...Houston... :smiley: 

> Let's for now assume that the number of attributes (the number of Dictionary entries) doesn't rise much. Sorry for confusing the discussion with auto conditions (which are not that much related).

No problem.

> I'm not sure if the additional memory usage also has some negative performance effects (like having the dictionaries no longer fit in processor caches).

Well, that's mitigated by the fact that a cache line is 64bytes and by the fact there are three indirections here and cache is already "polluted". Even with this PR (which store the key alongside it's value, cleaning the caches) the improvement is big but not so much: the search still linear in _memory_.

> Nice ideas, we might even want to split the Attributes(Dictionary) into "commonly used attributes" and "not commonly used attributes", where we optimize for the commonly used attributes, at the expense of the less commonly used attributes becoming slower. SBO is Small Buffer Optimization? Also looks like a nice feature.

Splitting them is good, so searches are limited. However I though about having a `std::array<hash, value>` in `Dictionary` and a (gigantic) switch when you search one of those (so skipping linear search and picking values locally).
But I don't know how (or even if) it can be implemented with `char *`...

> that we get issues not at compile time, but at runtime at a PC of one of our content-writers that just doesn't know anything about C++ and string-hashing.

...This is different than _loading time_; or it's possible to 'attach' properties to Outfit which are read and cheched
as if they are variable after loading? I meant, really? ...You can inject a new attribute in an Outfit at runtime? :astonished: 

However, my fault: no "compile time checks" for loaded data (in C++11 I fear at best you can check syntax).
But _before_ runtime, you have loading and tests. There's something content creators are required to check?

> Your active participation in the discussion is very valuable, please keep voicing your opinion and your ideas.

Thank's

petervdmeer commented:
> ...This is different than _loading time_; or it's possible to 'attach' properties to Outfit which are read and cheched
> as if they are variable after loading? I meant, really? ...You can inject a new attribute in an Outfit at runtime? astonished

I guess you already found out by yourself that the answer is `Yes, Outfits can get new attributes/properties at runtime, and also from plugins.` :)
I need to re-review this PR.

petervdmeer commented on source/Dictionary.cpp:
```diff
@@ -60,6 +37,48 @@ namespace {
+
+// The dataset is sorted by its 'char *' key (which is indeed the only
+// allowed key used when adding a new element because it's unique), so
+// a bynary search is not possible: do a 'find'
```
bynary -> binary

mike-f1 replied:
Done.

petervdmeer commented:
> @petervdmeer , now only the attributes which are set in code using the overloaded `_fnv1a` (which is still to be changed, as per previous review request) are using the hashing function: the others are still doing a string comparison.

Sorry if my previous comment wasn't clear; only the attributes that are used _in code_ are actually used _in flight_. The other attributes might exist on an outfit, but they are only relevant when buying and selling Outfits off and on ships, so the string compared attributes are not performance critical. (Since there is no "realtime flight code that can deal with them".)

> Cons: the `vector` is sorted by strings, and this means that a search by hash can be done only with a linear search.

The "realtime flight performance critical" attributes don't need to be strings, they can be hashes just fine.
The "non-realtime" attributes don't need to have high-performance properties.

> Pros: ...the linear search, given the data size, still faster than a binary search.

Nice, but I guess we can optimize for the realtime properties that are used _every frame, for every active ship_ and can accept a less performant option for properties that are only used during loading and buying/selling of outfits.

mike-f1 commented:
> Nice, but I guess we can optimize for the realtime properties that are used every frame, for every active ship and can accept a less performant option for properties that are only used during loading and buying/selling of outfits.

...I tried, but (at the moment?) I don't see a way it can be done: how actual code can differentiate between "realtime flight performance critical" (used in code) properties and "non-realtime" properties (not used in code)?

Anyway, the last update contains that try in a couple of new commits.

petervdmeer commented:
> ...I tried, but (at the moment?) I don't see a way it can be done: how actual code can differentiate between "realtime flight performance critical" (used in code) properties and "non-realtime" properties (not used in code)?

I created issue https://github.com/endless-sky/endless-sky/issues/8710 where I tried to describe this difference and what I feel needs to change to achieve it.
The "non-realtime" properties are also used in code, but only at the outfitter and during plundering. (Actions that are not performed every frame, thus non-realtime.)

I'm also closing this PR; when we have the list of "realtime flight performance critical" properties split off from the non-performance critical attributes, then I expect that we can do better than both string indexing or hashing because the full list will be known at compile time. (Then we can do something along the line of vector(integer)-based indexing as done in https://github.com/endless-sky/endless-sky/pull/7685 .)
